swap/protocol_test.go

```diff
@@ -44,68 +45,174 @@ func init() {
+	if err == nil {
+		return fmt.Errorf("Unexpected peer disconnect")
+	}
+
+	if err.Error() != "timed out waiting for peers to disconnect" {
+		return err
+	}
+
+	return nil
+}
+
+// creates a new HandshakeMsg
+func newSwapHandshakeMsg(contractAddress common.Address, chainID uint64) *HandshakeMsg {
+	return &HandshakeMsg{
+		ContractAddress: contractAddress,
+		ChainID:         chainID,
+	}
+}
+
+// creates the correct HandshakeMsg based on Swap instance
+func correctSwapHandshakeMsg(swap *Swap) *HandshakeMsg {
+	return newSwapHandshakeMsg(swap.GetParams().ContractAddress, swap.chainID)
+}
+
+// TestHandshake tests the correct handshake scenario
+func TestHandshake(t *testing.T) {
+	// setup the protocolTester, which will allow protocol testing by sending messages
+	protocolTester, clean, err := newSwapTester(t)
+	defer clean()
 	if err != nil {
 		t.Fatal(err)
 	}
 
-	// run the exchange:
-	// trigger a `EmitChequeMsg`
-	// expect HandshakeMsg on each node
-	err = protocolTester.TestExchanges(p2ptest.Exchange{
-		Label: "TestHandshake",
-		Triggers: []p2ptest.Trigger{
-			{
-				Code: 0,
-				Msg: &HandshakeMsg{
-					ContractAddress: swap.GetParams().ContractAddress,
-				},
-				Peer: creditor.ID(),
-			},
+	err = protocolTester.testHandshake(
+		correctSwapHandshakeMsg(protocolTester.swap),
+		correctSwapHandshakeMsg(protocolTester.swap),
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+// TestHandshakeInvalidChainID tests that a handshake with the wrong chain id is rejected
+func TestHandshakeInvalidChainID(t *testing.T) {
+	// setup the protocolTester, which will allow protocol testing by sending messages
+	protocolTester, clean, err := newSwapTester(t)
+	defer clean()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = protocolTester.testHandshake(
+		correctSwapHandshakeMsg(protocolTester.swap),
+		newSwapHandshakeMsg(protocolTester.swap.GetParams().ContractAddress, 1234),
+		&p2ptest.Disconnect{
+			Peer:  protocolTester.Nodes[0].ID(),
+			Error: errors.New("Handshake error: Message handler error: (msg code 0): different chain id"),
 		},
-		Expects: []p2ptest.Expect{
-			{
-				Code: 0,
-				Msg: &HandshakeMsg{
-					ContractAddress: swap.GetParams().ContractAddress,
-				},
-				Peer: debitor.ID(),
-			},
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+// TestHandshakeEmptyContract tests that a handshake with an empty contract address is rejected
+func TestHandshakeEmptyContract(t *testing.T) {
+	// setup the protocolTester, which will allow protocol testing by sending messages
+	protocolTester, clean, err := newSwapTester(t)
+	defer clean()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = protocolTester.testHandshake(
+		correctSwapHandshakeMsg(protocolTester.swap),
+		newSwapHandshakeMsg(common.Address{}, 1234),
+		&p2ptest.Disconnect{
+			Peer:  protocolTester.Nodes[0].ID(),
+			Error: errors.New("Handshake error: Message handler error: (msg code 0): empty address in handshake"),
```

Comment:
Can you somehow create this error by including the (exported) error messages from the different protocols? Such that when if the content of the error message is changed in one protocol, it does not make our test fail?

For example, the expected error message here is a concatenation of:
`protocols.ErrHandshake` + `protocols.ErrHandler` + `swap.ErrDifferentChainID`.

Reply:
done